README.md
# ✨ kumi_popup_window

[![pub package](https://img.shields.io/pub/v/kumi_popup_window.svg)](https://pub.dev/packages/kumi_popup_window)

**[Language]** [English](README.md) | [中文文档](README_CN.md)

A simple and easy-to-use pop-up window. There are no flashy preset templates. Everything you want to pop up is customized by you.

**1.Install**

```yaml
dependencies:
  kumi_popup_window: ^2.0.1
```

**2.Import**

```dart
import 'package:kumi_popup_window/kumi_popup_window.dart';
```

**3.Use**
```dart
//both sowPopupWindow and createPopupWindow
showPopupWindow(
  context,
  gravity: KumiPopupGravity.rightBottom,
  //curve: Curves.elasticOut,
  bgColor: Colors.grey.withOpacity(0.5),
  clickOutDismiss: true,
  clickBackDismiss: true,
  customAnimation: false,
  customPop: false,
  customPage: false,
  //targetRenderBox: (btnKey.currentContext.findRenderObject() as RenderBox),
  //needSafeDisplay: true,
  underStatusBar: false,
  underAppBar: true,
  offsetX: 0,
  offsetY: 0,
  duration: Duration(milliseconds: 200),
  onShowStart: (pop) {
    print("showStart");
  },
  onShowFinish: (pop) {
    print("showFinish");
  },
  onDismissStart: (pop) {
    print("dismissStart");
  },
  onDismissFinish: (pop) {
    print("dismissFinish");
  },
  onClickOut: (pop){
    print("onClickOut");
  },
  onClickBack: (pop){
    print("onClickBack");
  },
  childFun: (pop) {
    return Container(
      key: GlobalKey(),
      padding: EdgeInsets.all(10),
      height: 100,
      width: 100,
      color: Colors.redAccent,
    );
  },
);

```

**4.Demo screen**

<img src="demo/popup_1.png" height="400" alt="Screenshot"/> <img src="demo/popup_2.png" height="400" alt="Screenshot"/>

**note**
**Data update if using setState, please see [example] (example/lib/main.dart)**

Brief description, see [here](lib/kumi_popup_window.dart) for details


params|description|default
--|--|--|
childFun|Method for customizing the content of the popup window. The returned widget will be used as the content of the popup window|null，(require)
gravity|Relative position of the pop-up window.|KumiPopupGravity.center
curve|Animation interpolator|Curves.decelerate
customAnimation|Custom popupWindow animation|false
customPop|Customize the position and animation of popupWindow|false
customPage|Customize the entire page, including Scaffold|false
bgColor|The color of the mask layer|Colors.black.withOpacity(0.5)
targetRenderBox|RenderBox of the target widget|null
underStatusBar|When the top pops up, is it below the statusBar|false
underAppBar|When the top pops up, is it below the AppBar|false
clickOutDismiss|Click outside the pop-up window to dismiss|true
clickBackDismiss|Click the physical back button to dismiss|true
offsetX|Offset at horizontal axis edge|0.0
offsetY|Offset at vertical axis edge|0.0
duration|Duration of the animation|Duration(milliseconds: 300)
needSafeDisplay|Do you need to display pop ups safely|false
onShowStart|When the popupWindow show animation start|null
onShowEnd|When the popupWindow show animation finish|null
onDismissStart|When the popupWindow dismiss animation start|null
onDismissEnd|When the popupWindow dismiss animation finish|null
onClickOut|Click on the listener outside the popup|null
onClickBack|Listening on clicking the physical back button|null

property/method|description
--|--|
child|Popup window widget,Get its size and position by GlobalKey
isShow|Is the current pop-up displayed
controller|Animation Controller
show()|popup window show
dismiss()|popup window dismiss

## About

* QQ：876625596
* Email：876625596@qq.com
* Github：https://github.com/q876625596

## License

MIT License

Copyright (c) 2020 q876625596

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
## [2.0.1]
* 更新flutter到3.0.0
* Update flutter to 3.0.0,


## [2.0.0-nullsafety]
* 更新flutter到2.0.2，空安全支持
* Update flutter to 2.0.2, null safety support

## [1.0.12]
* 新增弹出框的安全显示控制参数needSafeDisplay，详见注释
* 优化相对于屏幕的弹出框，如果是相对于屏幕的弹框，可以不写宽高，默认可撑满屏幕
* Newly added security display control parameter needsafedisplay in pop-up box, see note for details
* Optimize the pop-up box relative to the screen. If it is a pop-up box relative to the screen, you can not write the width and height. By default, it can hold the full screen

## [1.0.11]
* 测试升级到flutter 1.22.2版本成功
* 优化动画，新增curve属性用于自定义插值器
* Flutter upgrade to 1.22.2 test success
* Optimize animation, add curve attribute for custom interpolator

## [1.0.10]
* 测试升级到flutter 1.7.3版本成功
* Flutter upgrade to 1.7.3 test success

## [1.0.9]
* 调用dismiss()之前判断isShow
* Determine isShow before calling dismiss ()

## [1.0.8]
* 删除SafeArea
* Delete SafeArea

## [1.0.7]
* 新增isShow属性
* Added isShow property

## [1.0.6]
* dismiss()方法添加禁止动画的参数
* dismiss () method adds a parameter that prohibits animation

## [1.0.5]
* 修改默认遮罩层颜色
* Modify the default mask layer color

## [1.0.4]
* 为dismiss()法新增可选参数
* Added optional parameters for dismiss()

## [1.0.3]
* fix bug

## [1.0.2]

* 新增onClickOut()和onClickBack()
* Added onClickOut () and onClickBack ()

## [1.0.1]

*修改部分方法
* Modification method.

## [1.0.0]

* first version.




lib/PopRoute.dart
import 'package:flutter/material.dart';

class PopRoute extends PopupRoute {
  final Widget _child;

  PopRoute({
    required Widget child,
  }):
        _child = child;


  @override
  Color? get barrierColor => null;

  @override
  bool get barrierDismissible => true;

  @override
  String? get barrierLabel => null;

  @override
  Duration get transitionDuration => Duration(milliseconds: 0);
  
  @override
  Widget buildPage(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
    return _child;
  }

  @override
  Widget buildTransitions (BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    return child;
    //return FadeTransition(opacity: animation, child: child);
    /*if (child is AnimatedWidget) {
      return child;
    }
    const Offset topLeft = const Offset(0.0, 0.0);
    const Offset topRight = const Offset(0.0, 0.0);
    const Offset bottomLeft = const Offset(0.0, 1.0);
    const Offset bottomRight = const Offset(0.0, 0.0);
    Offset startOffset = bottomLeft;
    Offset endOffset = topLeft;

      startOffset = bottomLeft;
      endOffset = bottomRight;

    return SlideTransition(
      position: Tween<Offset>(
        begin: startOffset,
        end: endOffset,
      ).animate(animation),
      child: child,
    );*/
  }



}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:kumi_popup_window/kumi_popup_window.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  GlobalKey btnKey = GlobalKey();
  KumiPopupWindow? popupWindow;

  ValueNotifier<bool> isSelect = ValueNotifier(false);
  var aaa = "false";

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title ?? ""),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            ValueListenableBuilder(
                valueListenable: isSelect,
                builder: (context, bool select, child) {
                  return Text(isSelect.value == false ? "popup child function onclick is false" : "popup child function onclick is true");
                }),
            SizedBox(
              height: 20,
            ),
            MaterialButton(
                key: btnKey,
                height: 50,
                child: Text("popup"),
                color: Colors.redAccent,
                onPressed: () {
                  /*showModalBottomSheet(
                      context: context,
                      isScrollControlled: true,
                      builder: (BuildContext context) {
                        return StatefulBuilder(
                          builder: (context1, StateSetter setBottomSheetState) {
                            return SingleChildScrollView(
                              child: Container(
                                padding: EdgeInsets.only(bottom: MediaQuery.of(context).viewInsets.bottom),
                                child: Column(
                                  mainAxisSize: MainAxisSize.min,
                                  children: [
                                    TextButton(
                                      onPressed: () {
                                        setBottomSheetState(() {
                                        aaa = "true";
                                      });
                                      },
                                      child: Text("asdasdasd"),
                                    ),
                                    TextField(
                                      controller: TextEditingController(text: aaa),
                                      keyboardType: TextInputType.number,
                                      inputFormatters: [WhitelistingTextInputFormatter.digitsOnly], //只允许输入数字
                                      textInputAction: TextInputAction.done,
                                    ),
                                  ],
                                ),
                              ),
                            );
                          },
                        );
                      });*/
                  showPopupWindow(
                    context,
                    //childSize:Size(240, 800),
                    gravity: KumiPopupGravity.centerBottom,
                    //curve: Curves.elasticOut,
                    bgColor: Colors.grey.withOpacity(0.5),
                    clickOutDismiss: true,
                    clickBackDismiss: true,
                    customAnimation: false,
                    customPop: false,
                    customPage: false,
                    // targetRenderBox: (btnKey.currentContext.findRenderObject() as RenderBox),
                    //needSafeDisplay: true,
                    underStatusBar: false,
                    underAppBar: false,
                    //offsetX: -180,
                    //offsetY: 50,
                    duration: Duration(milliseconds: 300),
                    onShowStart: (pop) {
                      print("showStart");
                    },
                    onShowFinish: (pop) {
                      print("showFinish");
                    },
                    onDismissStart: (pop) {
                      print("dismissStart");
                    },
                    onDismissFinish: (pop) {
                      print("dismissFinish");
                    },
                    onClickOut: (pop) {
                      print("onClickOut");
                    },
                    onClickBack: (pop) {
                      print("onClickBack");
                    },
                    childFun: (pop) {
                      return StatefulBuilder(
                          key: GlobalKey(),
                          builder: (popContext, popState) {
                            return GestureDetector(
                              onTap: () {
                                //isSelect.value = !isSelect.value;
                                popState(() {
                                  aaa = "sasdasd";
                                });
                              },
                              child: Container(
                                padding: EdgeInsets.all(10),
                                height: 800,
                                width: 300,
                                color: Colors.redAccent,
                                alignment: Alignment.center,
                                child: Text(aaa),
                              ),
                            );
                          });
                    },
                  );
                }),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
